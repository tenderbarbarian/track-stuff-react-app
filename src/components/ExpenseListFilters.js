import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			calendarFocused: null
		};
	}
	onDatesChange = ({ startDate, endDate }) => {
		//this.props.dispatch(setStartDate(startDate));
		//this.props.dispatch(setEndDate(endDate));
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};
	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};
	onTextChange = (e) => {
		//this.props.dispatch(setTextFilter(e.target.value));
		this.props.setTextFilter(e.target.value);
	};
	onSortChange = (e) => {
		if (e.target.value === 'amount') {
			//this.props.dispatch(sortByAmount());
			this.props.sortByAmount();
		} else if (e.target.value === 'date') {
			//this.props.dispatch(sortByDate());
			this.props.sortByDate();
		}
	};
	render() {
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						<input
							type="text"
							className="text-input"
							value={this.props.filters.text}
							onChange={this.onTextChange}
							placeholder="Search Expenses"
						/>
					</div>
					<div className="input-group__item">
						<select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
							<option value="date">Date</option>
							<option value="amount">Amount</option>
						</select>
					</div>
					<div className="input-group__item">
						<DateRangePicker
							startDate={this.props.filters.startDate}
							startDateId="your_unique_start_date_id"
							endDate={this.props.filters.endDate}
							endDateId="your_unique_end_date_id"
							onDatesChange={this.onDatesChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							showClearDates={true}
							numberOfMonths={1}
							isOutsideRange={() => false}
						/>
					</div>
				</div>
			</div>
		);
	}
}

// const ExpenseListFilters = (props) => (
//     <div>
//         <input
//             type='text'
//             value={props.filters.text}
//             onChange={(e)=>{
//                 props.dispatch(setTextFilter(e.target.value));
//             }}
//         />
//         <select
//             value={props.filters.sortBy}
//             onChange={(e)=>{
//                 if (e.target.value === 'amount') {
//                     props.dispatch(sortByAmount());
//                     } else if (e.target.value === 'date') {
//                         props.dispatch(sortByDate());
//                         };
//             }}
//         >
//             <option value="date">Date</option>
//             <option value="amount">Amount</option>
//         </select>
//     </div>
// );

const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
};

const mapDispatchToProps = (dispatch) => ({
	sortByAmount: () => dispatch(sortByAmount()),
	sortByDate: () => dispatch(sortByDate()),
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setStartDate: (startDate) => dispatch(setStartDate(startDate))
});

const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;
