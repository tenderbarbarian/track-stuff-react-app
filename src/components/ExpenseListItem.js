import React from 'react';
import {connect} from 'react-redux';
import {Link}  from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseListItem = ({id, description, amount, createdAt, dispatch})=>(
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{numeral(amount).format('$0,0.00')} 
        - 
        {moment(createdAt).format('MMMM Do, YYYY')}</p> 
    </div>
);


const ConnectedExpenseListItem = connect()(ExpenseListItem);

export default ConnectedExpenseListItem;