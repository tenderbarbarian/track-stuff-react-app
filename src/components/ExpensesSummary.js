import React from 'react';
import { connect } from 'react-redux';
import expensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

export const ExpensesSummary = (props) => (
	<div className="page-header">
		<div className="content-container">
			<h1 className="page-header__title">
				Viewing <span>{props.expenses.length}</span> expense{!(props.expenses.length === 1) && 's'} totalling:{' '}
				<span>{numeral(expensesTotal(props.expenses)).format('$0,0.00')}</span>
			</h1>
			<div className="page-header__actions">
				<Link className="button" to="/create">
					Add expense
				</Link>
			</div>
		</div>
	</div>
);

const mapStateToProps = (state) => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};

const ConnectedExpensesSummary = connect(mapStateToProps)(ExpensesSummary);
export default ConnectedExpensesSummary;
