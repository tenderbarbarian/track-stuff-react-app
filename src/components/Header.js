import React from 'react';
import {NavLink} from 'react-router-dom';

const Header =()=>(
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName="is-active" to="/" exact={true}>Home</NavLink>
        <NavLink activeClassName="is-active" to="/create">Create new</NavLink>
    </header>
);

export default Header;