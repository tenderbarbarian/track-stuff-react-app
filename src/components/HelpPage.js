import React from 'react';


const HelpPage = ()=> (
    <div>
        HElp page 
    </div>
);

export default HelpPage;