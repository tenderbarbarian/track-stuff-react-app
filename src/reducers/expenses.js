

const expensesReducer = (state = [], action)=>{
    switch (action.type){
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id })=> (id !== action.id));
        case 'EDIT_EXPENSE':1
            return state.map((expenseItem)=> {
                if(expenseItem.id === action.id){
                    return{
                        ...expenseItem,
                        ...action.updates
                    };
                } else{
                    return expenseItem;
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;