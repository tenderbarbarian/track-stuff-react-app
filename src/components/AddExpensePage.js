import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import { history } from '../routers/AppRouter';

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		//props.dispatch(startAddExpense(expense));
		this.props.startAddExpense(expense); //used with mapDispatchTo props. Doing it this way so that it's easier to test
		console.log('redirect');
		this.props.history.push('/dashboard'); //go to the home page
	};
	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Add Expense</h1>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm onSubmit={this.onSubmit} />
				</div>
			</div>
		);
	}
}

// const AddExpensePage = (props)=> (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit= {(expense)=>{
//                 props.dispatch(addExpense(expense));
//                 //props.onSubmit(expense);  //used with mapDispatchTo props. Doing it this way so that it's easier to test
//                 props.history.push('/');  //go to the home page
//             }}
//         />
//     </div>
// );

//works great as here:
//export default connect()(AddExpensePage);

//changing stuff for testing:
//build in Redux function mapDispatchToprops
const mapDispatchToProps = (dispatch) => {
	return {
		startAddExpense: (expense) => dispatch(startAddExpense(expense))
	};
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
