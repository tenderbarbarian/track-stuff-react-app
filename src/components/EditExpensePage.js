import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { history } from '../routers/AppRouter';

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => {
		//this.props.dispatch(startEditExpense(this.props.expense.id, expense));
		this.props.startEditExpense(this.props.expense.id, expense);
		this.props.history.push('/dashboard'); //go to the home page
	};
	onRemove = (e) => {
		//this.props.dispatch(removeExpense({id: this.props.match.params.id}));
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/dashboard');
	};
	render() {
		return (
			<div>
				<div className="page-header">
					<h1 className="page-header__title">Edit Expense</h1>
				</div>
				<div className="content-container">
					{/* This is my edit expense. ID: {this.props.match.params.id} */}
					<ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
					<button onClick={this.onRemove} className="button button--grey">
						Remove Expense
					</button>
				</div>
			</div>
		);
	}
}

// const EditExpensePage = (props)=> {
//     // console.log(props);
//     return (
//         <div>
//             {/* This is my edit expense. ID: {props.match.params.id} */}
//             <ExpenseForm
//             expense = {props.expense}
//             onSubmit= {(expense)=>{
//                 props.dispatch(startEditExpense(props.expense.id, expense));
//                 props.history.push('/');  //go to the home page
//             }}
//             />
//             <button onClick={(e)=>{
//                 props.dispatch(removeExpense({id: props.match.params.id}));
//                 props.history.push('/');
//                 }}>Remove
//             </button>
//         </div>
//     )
// };

const mapDispatchToProps = (dispatch) => {
	return {
		startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
		startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
	};
};

const mapStateToProps = (state, props) => {
	return {
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	};
};

//export default connect(mapStateToProps)(EditExpensePage);
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
