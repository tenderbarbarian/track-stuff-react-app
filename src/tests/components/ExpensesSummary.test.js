import React from 'react';
import expenses from '../fixtures/expenses';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('should render ExpensesSummary component corretly if there are some expenses', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary component corretly if there are some expenses', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot();
});