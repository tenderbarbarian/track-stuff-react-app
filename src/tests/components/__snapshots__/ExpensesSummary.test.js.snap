// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary component corretly if there are some expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       expense
      s
       totalling:
       
      <span>
        $10,174.99
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary component corretly if there are some expenses 2`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense
       totalling:
       
      <span>
        $43.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;
